File: clients/client-python/tests/unittests/test_supports_jobs.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
import unittest
from http.client import HTTPResponse
from unittest.mock import Mock, patch

from gravitino import GravitinoClient
from gravitino.api.job.job_handle import JobHandle
from gravitino.api.job.job_template import JobType
from gravitino.api.job.shell_job_template import ShellJobTemplate
from gravitino.api.job.spark_job_template import SparkJobTemplate
from gravitino.dto.audit_dto import AuditDTO
from gravitino.dto.job.job_dto import JobDTO
from gravitino.dto.job.shell_job_template_dto import ShellJobTemplateDTO
from gravitino.dto.job.spark_job_template_dto import SparkJobTemplateDTO
from gravitino.dto.responses.base_response import BaseResponse
from gravitino.dto.responses.drop_response import DropResponse
from gravitino.dto.responses.job_list_response import JobListResponse
from gravitino.dto.responses.job_response import JobResponse
from gravitino.dto.responses.job_template_list_response import JobTemplateListResponse
from gravitino.dto.responses.job_template_response import JobTemplateResponse
from gravitino.utils import Response
from tests.unittests import mock_base


@mock_base.mock_data
class TestSupportsJobs(unittest.TestCase):
    _metalake_name = "metalake_demo"

    def test_list_job_templates(self, *mock_methods):

        gravitino_client = GravitinoClient(
            uri="http://localhost:8090",
            metalake_name=self._metalake_name,
        )

        templates = [self._new_shell_job_template(), self._new_spark_job_template()]
        template_dtos = [
            self._new_shell_job_template_dto(templates[0]),
            self._new_spark_job_template_dto(templates[1]),
        ]
        job_template_list_resp = JobTemplateListResponse(
            _job_templates=template_dtos, _code=0
        )
        json_str = job_template_list_resp.to_json()
        mock_resp = self._mock_http_response(json_str)

        with patch(
            "gravitino.utils.http_client.HTTPClient.get", return_value=mock_resp
        ):
            result_templates = gravitino_client.list_job_templates()
            self.assertEqual(templates, result_templates)

        # test with empty response
        job_template_list_resp = JobTemplateListResponse(_job_templates=[], _code=0)
        json_str = job_template_list_resp.to_json()
        mock_resp = self._mock_http_response(json_str)

        with patch(
            "gravitino.utils.http_client.HTTPClient.get", return_value=mock_resp
        ):
            result_templates = gravitino_client.list_job_templates()
            self.assertEqual([], result_templates)

    def test_register_job_template(self, *mock_methods):
        gravitino_client = GravitinoClient(
            uri="http://localhost:8090",
            metalake_name=self._metalake_name,
        )

        shell_template = self._new_shell_job_template()
        resp = BaseResponse(_code=0)
        mock_resp = self._mock_http_response(resp.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.post", return_value=mock_resp
        ):
            gravitino_client.register_job_template(shell_template)

    def test_get_job_template(self, *mock_methods):
        gravitino_client = GravitinoClient(
            uri="http://localhost:8090",
            metalake_name=self._metalake_name,
        )

        shell_template = self._new_shell_job_template()
        shell_template_dto = self._new_shell_job_template_dto(shell_template)
        resp = JobTemplateResponse(_job_template=shell_template_dto, _code=0)
        mock_resp = self._mock_http_response(resp.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.get", return_value=mock_resp
        ):
            result_template = gravitino_client.get_job_template(shell_template.name)
            self.assertEqual(shell_template, result_template)

    def test_delete_job_template(self, *mock_methods):
        gravitino_client = GravitinoClient(
            uri="http://localhost:8090",
            metalake_name=self._metalake_name,
        )

        shell_template = self._new_shell_job_template()
        resp = DropResponse(_dropped=True, _code=0)
        mock_resp = self._mock_http_response(resp.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.delete", return_value=mock_resp
        ):
            result = gravitino_client.delete_job_template(shell_template.name)
            self.assertTrue(result)

        resp = DropResponse(_dropped=False, _code=0)
        mock_resp = self._mock_http_response(resp.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.delete", return_value=mock_resp
        ):
            result = gravitino_client.delete_job_template(shell_template.name)
            self.assertFalse(result)

    def test_list_jobs(self, *mock_methods):
        gravitino_client = GravitinoClient(
            uri="http://localhost:8090",
            metalake_name=self._metalake_name,
        )

        job_template_name = "test_shell_job"
        job_dtos = [
            self._new_job_dto(job_template_name),
            self._new_job_dto(job_template_name),
        ]
        resp = JobListResponse(_jobs=job_dtos, _code=0)
        mock_resp = self._mock_http_response(resp.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.get", return_value=mock_resp
        ):
            result_jobs = gravitino_client.list_jobs(job_template_name)
            self.assertEqual(len(job_dtos), len(result_jobs))
            for i, job_dto in enumerate(job_dtos):
                self._compare_job_handle(result_jobs[i], job_dto)

    def test_run_job(self, *mock_methods):
        gravitino_client = GravitinoClient(
            uri="http://localhost:8090",
            metalake_name=self._metalake_name,
        )

        job_template_name = "test_shell_job"
        job_dto = self._new_job_dto(job_template_name)
        resp = JobResponse(_job=job_dto, _code=0)
        mock_resp = self._mock_http_response(resp.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.post", return_value=mock_resp
        ):
            job_handle = gravitino_client.run_job(job_template_name, {})
            self._compare_job_handle(job_handle, job_dto)

    def test_get_job(self, *mock_methods):
        gravitino_client = GravitinoClient(
            uri="http://localhost:8090",
            metalake_name=self._metalake_name,
        )

        job_template_name = "test_shell_job"
        job_dto = self._new_job_dto(job_template_name)
        resp = JobResponse(_job=job_dto, _code=0)
        mock_resp = self._mock_http_response(resp.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.get", return_value=mock_resp
        ):
            job_handle = gravitino_client.get_job(job_dto.job_id())
            self._compare_job_handle(job_handle, job_dto)

        # test with invalid input
        with self.assertRaises(ValueError):
            gravitino_client.get_job("")

        with self.assertRaises(ValueError):
            gravitino_client.get_job(None)

    def test_cancel_job(self, *mock_methods):
        gravitino_client = GravitinoClient(
            uri="http://localhost:8090",
            metalake_name=self._metalake_name,
        )

        job_template_name = "test_shell_job"
        job_dto = self._new_job_dto(job_template_name)
        resp = JobResponse(_job=job_dto, _code=0)
        mock_resp = self._mock_http_response(resp.to_json())

        with patch(
            "gravitino.utils.http_client.HTTPClient.post", return_value=mock_resp
        ):
            job_handle = gravitino_client.cancel_job(job_dto.job_id())
            self._compare_job_handle(job_handle, job_dto)

        # test with invalid input
        with self.assertRaises(ValueError):
            gravitino_client.cancel_job("")

        with self.assertRaises(ValueError):
            gravitino_client.cancel_job(None)

    def _new_shell_job_template(self):
        return (
            ShellJobTemplate.builder()
            .with_name("test_shell_job")
            .with_comment("This is a test shell job template")
            .with_executable("/path/to/executable.sh")
            .with_arguments(["arg1", "arg2"])
            .with_environments({"ENV_VAR": "value"})
            .with_custom_fields({"custom_field": "custom_value"})
            .with_scripts(["/path/to/script1.sh", "/path/to/script2.sh"])
            .build()
        )

    def _new_spark_job_template(self):
        return (
            SparkJobTemplate.builder()
            .with_name("test_spark_job")
            .with_comment("This is a test spark job template")
            .with_executable("/path/to/spark-demo.jar")
            .with_class_name("com.example.SparkJob")
            .with_jars(["/path/to/jar1.jar", "/path/to/jar2.jar"])
            .with_files(["/path/to/file1.txt", "/path/to/file2.txt"])
            .with_archives(["/path/to/archive1.zip", "/path/to/archive2.zip"])
            .with_configs({"spark.executor.memory": "2g", "spark.driver.memory": "1g"})
            .build()
        )

    def _new_shell_job_template_dto(self, template: ShellJobTemplate):
        return ShellJobTemplateDTO(
            _job_type=JobType.SHELL,
            _name=template.name,
            _comment=template.comment,
            _executable=template.executable,
            _arguments=template.arguments,
            _environments=template.environments,
            _custom_fields=template.custom_fields,
            _scripts=template.scripts,
        )

    def _new_spark_job_template_dto(self, template: SparkJobTemplate):
        return SparkJobTemplateDTO(
            _job_type=JobType.SPARK,
            _name=template.name,
            _comment=template.comment,
            _executable=template.executable,
            _arguments=template.arguments,
            _environments=template.environments,
            _custom_fields=template.custom_fields,
            _class_name=template.class_name,
            _jars=template.jars,
            _files=template.files,
            _archives=template.archives,
            _configs=template.configs,
        )

    def _mock_http_response(self, json_str: str):
        mock_http_resp = Mock(HTTPResponse)
        mock_http_resp.getcode.return_value = 200
        mock_http_resp.read.return_value = json_str
        mock_http_resp.info.return_value = None
        mock_http_resp.url = None
        mock_resp = Response(mock_http_resp)
        return mock_resp

    def _new_job_dto(self, job_template_name: str) -> JobDTO:
        return JobDTO(
            _job_id="job-123",
            _job_template_name=job_template_name,
            _status=JobHandle.Status.QUEUED,
            _audit=AuditDTO(_creator="test", _create_time="2023-10-01T00:00:00Z"),
        )

    def _compare_job_handle(self, job_handle: JobHandle, job_dto: JobDTO):
        self.assertEqual(job_handle.job_id(), job_dto.job_id())
        self.assertEqual(job_handle.job_template_name(), job_dto.job_template_name())
        self.assertEqual(job_handle.job_status(), job_dto.status())

File: clients/client-java/src/test/java/org/apache/gravitino/client/TestSupportsJobs.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.client;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Lists;
import java.time.Instant;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import org.apache.gravitino.dto.AuditDTO;
import org.apache.gravitino.dto.job.JobDTO;
import org.apache.gravitino.dto.job.JobTemplateDTO;
import org.apache.gravitino.dto.job.ShellJobTemplateDTO;
import org.apache.gravitino.dto.job.SparkJobTemplateDTO;
import org.apache.gravitino.dto.requests.JobRunRequest;
import org.apache.gravitino.dto.requests.JobTemplateRegisterRequest;
import org.apache.gravitino.dto.responses.BaseResponse;
import org.apache.gravitino.dto.responses.DropResponse;
import org.apache.gravitino.dto.responses.ErrorResponse;
import org.apache.gravitino.dto.responses.JobListResponse;
import org.apache.gravitino.dto.responses.JobResponse;
import org.apache.gravitino.dto.responses.JobTemplateListResponse;
import org.apache.gravitino.dto.responses.JobTemplateResponse;
import org.apache.gravitino.exceptions.InUseException;
import org.apache.gravitino.exceptions.JobTemplateAlreadyExistsException;
import org.apache.gravitino.exceptions.MetalakeNotInUseException;
import org.apache.gravitino.exceptions.NoSuchJobException;
import org.apache.gravitino.exceptions.NoSuchJobTemplateException;
import org.apache.gravitino.exceptions.NoSuchMetalakeException;
import org.apache.gravitino.job.JobHandle;
import org.apache.gravitino.job.JobTemplate;
import org.apache.hc.core5.http.HttpStatus;
import org.apache.hc.core5.http.Method;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

public class TestSupportsJobs extends TestBase {

  private static final String METALAKE_FOR_JOB_TEST = "metalake_for_job_test";

  private static GravitinoMetalake metalake;

  @BeforeAll
  public static void setUp() throws Exception {
    TestBase.setUp();
    metalake = TestGravitinoMetalake.createMetalake(client, METALAKE_FOR_JOB_TEST);
  }

  @Test
  public void testListJobTemplates() throws JsonProcessingException {
    JobTemplateDTO template1 = newShellJobTemplateDTO("shell-job-template");
    JobTemplateDTO template2 = newSparkJobTemplateDTO("spark-job-template");
    List<JobTemplateDTO> templates = Lists.newArrayList(template1, template2);
    JobTemplateListResponse resp = new JobTemplateListResponse(templates);

    buildMockResource(Method.GET, jobTemplatesPath(), null, resp, HttpStatus.SC_OK);

    List<JobTemplate> jobTemplates = metalake.listJobTemplates();
    Assertions.assertEquals(2, jobTemplates.size());

    List<JobTemplate> expected =
        templates.stream()
            .map(org.apache.gravitino.dto.util.DTOConverters::fromDTO)
            .collect(Collectors.toList());

    Assertions.assertEquals(expected, jobTemplates);

    // Test throw NoSuchMetalakeException
    ErrorResponse errorResp =
        ErrorResponse.notFound(NoSuchMetalakeException.class.getSimpleName(), "mock error");
    buildMockResource(Method.GET, jobTemplatesPath(), null, errorResp, HttpStatus.SC_NOT_FOUND);
    Assertions.assertThrows(NoSuchMetalakeException.class, () -> metalake.listJobTemplates());

    // Test throw MetalakeNotInUseException
    ErrorResponse errorResp2 =
        ErrorResponse.notInUse(
            MetalakeNotInUseException.class.getSimpleName(),
            "mock error",
            new MetalakeNotInUseException("mock error"));
    buildMockResource(Method.GET, jobTemplatesPath(), null, errorResp2, HttpStatus.SC_CONFLICT);
    Assertions.assertThrows(MetalakeNotInUseException.class, () -> metalake.listJobTemplates());

    // Test throw RuntimeException
    ErrorResponse errorResp3 = ErrorResponse.internalError("mock error");
    buildMockResource(
        Method.GET, jobTemplatesPath(), null, errorResp3, HttpStatus.SC_INTERNAL_SERVER_ERROR);
    Assertions.assertThrows(RuntimeException.class, () -> metalake.listJobTemplates());
  }

  @Test
  public void testRegisterJobTemplate() throws JsonProcessingException {
    JobTemplateDTO templateDTO = newShellJobTemplateDTO("shell-job-template");
    JobTemplate template = org.apache.gravitino.dto.util.DTOConverters.fromDTO(templateDTO);
    JobTemplateRegisterRequest req = new JobTemplateRegisterRequest(templateDTO);

    BaseResponse resp = new BaseResponse();
    buildMockResource(Method.POST, jobTemplatesPath(), req, resp, HttpStatus.SC_OK);

    Assertions.assertDoesNotThrow(() -> metalake.registerJobTemplate(template));

    // Test throw JobTemplateAlreadyExistsException
    ErrorResponse errorResp =
        ErrorResponse.alreadyExists(
            JobTemplateAlreadyExistsException.class.getSimpleName(), "mock error");
    buildMockResource(Method.POST, jobTemplatesPath(), req, errorResp, HttpStatus.SC_CONFLICT);

    Assertions.assertThrows(
        JobTemplateAlreadyExistsException.class, () -> metalake.registerJobTemplate(template));
  }

  @Test
  public void testGetJobTemplate() throws JsonProcessingException {
    String jobTemplateName = "shell-job-template";
    JobTemplateDTO templateDTO = newShellJobTemplateDTO(jobTemplateName);
    JobTemplate expected = org.apache.gravitino.dto.util.DTOConverters.fromDTO(templateDTO);
    JobTemplateResponse resp = new JobTemplateResponse(templateDTO);

    buildMockResource(
        Method.GET, jobTemplatesPath() + "/" + jobTemplateName, null, resp, HttpStatus.SC_OK);

    JobTemplate actual = metalake.getJobTemplate(jobTemplateName);
    Assertions.assertEquals(expected, actual);

    // Test throw NoSuchJobTemplateException
    ErrorResponse errorResp =
        ErrorResponse.notFound(NoSuchJobTemplateException.class.getSimpleName(), "mock error");
    buildMockResource(
        Method.GET,
        jobTemplatesPath() + "/" + jobTemplateName,
        null,
        errorResp,
        HttpStatus.SC_NOT_FOUND);
    Assertions.assertThrows(
        NoSuchJobTemplateException.class, () -> metalake.getJobTemplate(jobTemplateName));
  }

  @Test
  public void testDeleteJobTemplate() throws JsonProcessingException {
    String jobTemplateName = "shell-job-template";
    DropResponse resp = new DropResponse(true);

    buildMockResource(
        Method.DELETE, jobTemplatesPath() + "/" + jobTemplateName, null, resp, HttpStatus.SC_OK);

    Assertions.assertTrue(metalake.deleteJobTemplate(jobTemplateName));

    // Test throw InUseException
    ErrorResponse errorResp2 =
        ErrorResponse.inUse(
            InUseException.class.getSimpleName(), "mock error", new InUseException("mock error"));
    buildMockResource(
        Method.DELETE,
        jobTemplatesPath() + "/" + jobTemplateName,
        null,
        errorResp2,
        HttpStatus.SC_CONFLICT);
    Assertions.assertThrows(
        InUseException.class, () -> metalake.deleteJobTemplate(jobTemplateName));
  }

  @Test
  public void testListJobs() throws JsonProcessingException {
    String jobTemplateName = "shell-job-template";
    String jobId1 = "job-1";
    String jobId2 = "job-2";

    List<JobDTO> jobs =
        Lists.newArrayList(newJobDTO(jobId1, jobTemplateName), newJobDTO(jobId2, jobTemplateName));

    JobListResponse resp = new JobListResponse(jobs);

    buildMockResource(Method.GET, jobRunsPath(), null, resp, HttpStatus.SC_OK);

    List<JobHandle> actualJobs = metalake.listJobs();
    Assertions.assertEquals(2, actualJobs.size());
    compare(jobs.get(0), actualJobs.get(0));
    compare(jobs.get(1), actualJobs.get(1));

    // Test throw NoSuchJobTemplateException
    ErrorResponse errorResp =
        ErrorResponse.notFound(NoSuchJobTemplateException.class.getSimpleName(), "mock error");
    buildMockResource(Method.GET, jobRunsPath(), null, errorResp, HttpStatus.SC_NOT_FOUND);
    Assertions.assertThrows(
        NoSuchJobTemplateException.class, () -> metalake.listJobs(jobTemplateName));

    // Test list jobs by job template name
    buildMockResource(
        Method.GET,
        jobRunsPath(),
        ImmutableMap.of("jobTemplateName", jobTemplateName),
        null,
        resp,
        HttpStatus.SC_OK);

    List<JobHandle> jobsByTemplate = metalake.listJobs(jobTemplateName);
    Assertions.assertEquals(2, jobsByTemplate.size());
    compare(jobs.get(0), jobsByTemplate.get(0));
    compare(jobs.get(1), jobsByTemplate.get(1));
  }

  @Test
  public void testGetJob() throws JsonProcessingException {
    String jobId = "job-1";
    String jobTemplateName = "shell-job-template";
    JobDTO expectedJob = newJobDTO(jobId, jobTemplateName);
    JobResponse resp = new JobResponse(expectedJob);

    buildMockResource(Method.GET, jobRunsPath() + "/" + jobId, null, resp, HttpStatus.SC_OK);

    JobHandle actualHandle = metalake.getJob(jobId);
    compare(expectedJob, actualHandle);

    // Test throw NoSuchJobTemplateException
    ErrorResponse errorResp =
        ErrorResponse.notFound(NoSuchJobException.class.getSimpleName(), "mock error");
    buildMockResource(
        Method.GET, jobRunsPath() + "/" + jobId, null, errorResp, HttpStatus.SC_NOT_FOUND);
    Assertions.assertThrows(NoSuchJobException.class, () -> metalake.getJob(jobId));
  }

  @Test
  public void testRunJob() throws JsonProcessingException {
    String jobTemplateName = "shell-job-template";
    String jobId = "job-1";
    JobDTO expectedJob = newJobDTO(jobId, jobTemplateName);
    JobResponse resp = new JobResponse(expectedJob);
    JobRunRequest req = new JobRunRequest(jobTemplateName, ImmutableMap.of());

    buildMockResource(Method.POST, jobRunsPath(), req, resp, HttpStatus.SC_OK);

    JobHandle actualHandle = metalake.runJob(jobTemplateName, ImmutableMap.of());
    compare(expectedJob, actualHandle);

    // Test throw NoSuchJobTemplateException
    ErrorResponse errorResp =
        ErrorResponse.notFound(NoSuchJobTemplateException.class.getSimpleName(), "mock error");
    buildMockResource(Method.POST, jobRunsPath(), req, errorResp, HttpStatus.SC_NOT_FOUND);
    Assertions.assertThrows(
        NoSuchJobTemplateException.class,
        () -> metalake.runJob(jobTemplateName, ImmutableMap.of()));
  }

  @Test
  public void testCancelJob() throws JsonProcessingException {
    String jobId = "job-1";
    String jobTemplateName = "shell-job-template";
    JobDTO expectedJob = newJobDTO(jobId, jobTemplateName);
    JobResponse resp = new JobResponse(expectedJob);

    buildMockResource(Method.POST, jobRunsPath() + "/" + jobId, null, resp, HttpStatus.SC_OK);

    JobHandle actualHandle = metalake.cancelJob(jobId);
    compare(expectedJob, actualHandle);

    // Test throw NoSuchJobException
    ErrorResponse errorResp =
        ErrorResponse.notFound(NoSuchJobException.class.getSimpleName(), "mock error");
    buildMockResource(
        Method.POST, jobRunsPath() + "/" + jobId, null, errorResp, HttpStatus.SC_NOT_FOUND);
    Assertions.assertThrows(NoSuchJobException.class, () -> metalake.cancelJob(jobId));
  }

  private void compare(JobDTO expected, JobHandle actual) {
    Assertions.assertEquals(expected.jobId(), actual.jobId());
    Assertions.assertEquals(expected.jobTemplateName(), actual.jobTemplateName());
    Assertions.assertEquals(expected.status(), actual.jobStatus());
  }

  private String jobTemplatesPath() {
    return "/api/metalakes/" + METALAKE_FOR_JOB_TEST + "/jobs/templates";
  }

  private String jobRunsPath() {
    return "/api/metalakes/" + METALAKE_FOR_JOB_TEST + "/jobs/runs";
  }

  private JobTemplateDTO newShellJobTemplateDTO(String name) {
    return ShellJobTemplateDTO.builder()
        .withJobType(JobTemplate.JobType.SHELL)
        .withName(name)
        .withComment("This is a shell job template")
        .withExecutable("/path/to/script.sh")
        .withArguments(Collections.emptyList())
        .withEnvironments(Collections.emptyMap())
        .withCustomFields(Collections.emptyMap())
        .withScripts(Collections.emptyList())
        .build();
  }

  private JobTemplateDTO newSparkJobTemplateDTO(String name) {
    return SparkJobTemplateDTO.builder()
        .withJobType(JobTemplate.JobType.SPARK)
        .withName(name)
        .withComment("This is a spark job template")
        .withExecutable("/path/to/spark-job.jar")
        .withArguments(Collections.emptyList())
        .withEnvironments(Collections.emptyMap())
        .withCustomFields(Collections.emptyMap())
        .withClassName("org.example.SparkJob")
        .withJars(Collections.emptyList())
        .withFiles(Collections.emptyList())
        .withArchives(Collections.emptyList())
        .withConfigs(Collections.emptyMap())
        .build();
  }

  private JobDTO newJobDTO(String jobId, String templateName) {
    return new JobDTO(
        jobId,
        templateName,
        JobHandle.Status.QUEUED,
        AuditDTO.builder().withCreator("test").withCreateTime(Instant.now()).build());
  }
}
